python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,204 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from abc import ABC, abstractmethod
+from typing import Dict, List, Optional, Set, Tuple, Union
+
+from iceberg.schema import Schema
+from iceberg.table.base import PartitionSpec, Table
+
+Identifier = Tuple[str, ...]
+Metadata = Dict[str, str]
+
+
+class Catalog(ABC):
+    """Base Catalog for table operations like - create, drop, load, list and others.
+
+    Attributes:
+        name(str): Name of the catalog
+        properties(Dict[str, str]): Catalog properties
+    """
+
+    def __init__(self, name: str, properties: Metadata):
```

Comment:
We could also generate this `__init__` very easily using the `@dataclass` by adding the decorator and defining the properties in the class itself:
```
    name: str
    properties: Metadata
```

This will also autogenerate the getters below:
```
    @property
    def name(self) -> str:
        return self._name

    @property
    def properties(self) -> Metadata:
        return self._properties
```

Reply:
@Fokko I am a little on the fence with having our interfaces (abstract classes) marked as `@dataclass`.  I'm not entirely sure if `ABC` and `@dataclass` play along nicely. Also, it forces all sub-classes to be annotated with `@dataclass` as well.